src/main/java/flink/graphs/Graph.java

```diff
@@ -1321,4 +1312,75 @@ public void remove() {
 							function.getClass(), 3, null, null);
 		}
 	}
+
+	public <MO> Graph<K, VV, EV> gsa (
+			MapFunction<Tuple3<Vertex<K, VV>, Edge<K, EV>, Vertex<K, VV>>, Tuple2<K, MO>> gather,
+			ReduceFunction<Tuple2<K, MO>> sum,
+			FlatJoinFunction<Tuple2<K, MO>, Vertex<K, VV>, Vertex<K, VV>> apply,
+			int maxIterations) {
+
+		DeltaIteration<Vertex<K, VV>, Vertex<K, VV>> iteration =
+				this.getVertices().iterateDelta(this.getVertices(), maxIterations, 0);
+
+		DataSet<Tuple3<Vertex<K, VV>, Edge<K, EV>, Vertex<K, VV>>> triplets = this
+				.getVertices()
```

Comment:
Retrieving the source vertices from the initial graph won't give you the updated vertex values, which will result into wrong behavior if your map-gather function uses the source vertex value. You can get the updated vertex values with `iteration.getSolutionSet()`.